_posts/2023-11-20-testing-quarkus-with-citrus.adoc

```diff
@@ -0,0 +1,506 @@
+synopsis: 'Explore how to verify Quarkus event-driven applications with the Citrus integration test framework.'
+author: christophd
+---
+:imagesdir: /assets/images/posts/citrus
+
+image::citrus-quarkus.png[Citrus & Quarkus,align="center"]
+
+This post shows how to combine Quarkus with the https://citrusframework.org[Citrus] test framework in order write automated tests for event-driven applications.
+https://citrusframework.org[Citrus] is an Open Source Java test framework focusing on messaging and integration testing in general.
+
+Developers can easily empower the *@QuarkusTest* with Citrus capabilities in order to produce and consume events during the test.
+As a result the test is able to interact with the Quarkus event-driven application by exchanging events and messages with real messaging communication.
+
+== Introducing the demo application
+
+In this post we use a Quarkus demo application called `food-market`.
+You can find the demo application code and all Citrus tests in https://github.com/citrusframework/citrus-samples/tree/main/demo/sample-quarkus[this GitHub code repository].
+The Quarkus application connects to Kafka streams as an event-driven application that produces and consumes various events (e.g. bookings, supplies, shipping events).
+The processed events and their individual status are stored in a PostgreSQL database.
+
+image::food-market-demo-application.png[Food Market App,align="center"]
+
+The food-market application matches incoming `booking` and `supply` events and produces `shipping` and `booking-completed` events accordingly.
+
+Each event references a product and specifies an amount as well as a price in a simple Json object structure.
+
+[source, json]
+----
+{ "client": "citrus-test", "product": "Pineapple", "amount":  100, "price":  0.99 }
+----
+
+Clients create the `booking` events and at the same time suppliers will add their individual `supply` events.
+The Quarkus food-market application consumes both event types and finds matching bookings and supplies.
+Once a booking and a supply do match in certain criteria the application produces `booking-completed` and `shipping` events as a result.
+
+Last but not least the booking client gets informed via email about the completed booking status.
+
+In a fully automated integration test we want to verify all events and their processing using real messaging communication with Kafka streams and database persistence.
+
+== Testing the application with Citrus
+
+The Quarkus application connects to different infrastructure (Kafka, PostgreSQL, Mail SMTP).
+The automated integration test should verify the message communication, the event processing and connectivity to all components.
+We will use the Citrus test framework as it provides the complete toolset for testing this kind of event-driven message-based solutions.
+
+The first thing to do is to add Citrus to the Quarkus project.
+The most convenient way is to import the `citrus-bom`.
```

Comment:
is the citrus-bom required?

how about getting the citrus-quarkus extension listed on code.quarkus.io? 

then hopefully quarkus ext add citrus-quarkus would be sufficient?

Reply:
https://github.com/quarkiverse/quarkiverse/wiki/Checklist#make-your-extension-available-in-the-tooling

Reply:
Or, better yet, https://hub.quarkiverse.io/checklist/#make-your-extension-available-in-the-tooling (same content as what @geoand shared, but formatted html... the indexing for that content is still a bit poor, but that's a different conversation!).

Reply:
+1 to list the extension on code.quarkus.io

the citrus-quarkus extension POM does not include all Citrus modules. How would the dependency resolving work in that case when there is no bom imported? 

In general the citrus-bom is not required, it only avoids to add the dependency version setting on all included `citrus-*` modules.

Reply:
Hmm. @aloubyansky, can you share thoughts on this?

Reply:
I wonder if the platform mechanism makes sense for this scenario: https://quarkus.io/guides/platform ... there would be a Citrus platform, which maps to the Citrus bom. Camel do this now, and there are a few smaller projects that also provide their own platform.

Reply:
I am not sure about adding this extension to the platform at this point.
The text makes sense to me. AFAIU, it's either a BOM or a shared version property.

Reply:
I was thinking more that we would create a new Citrus platform to hold its bom, not add it to the main Quarkus platform.

Reply:
I wouldn't do it either, so far we have only one platform in our registry.

Reply:
Oh, sorry, I'm using the wrong words. I think I mean platform member. 

Things like 
```
io.quarkus.platform:quarkus-kogito-bom-quarkus-platform-descriptor
io.quarkus.platform:quarkus-camel-bom-quarkus-platform-descriptor
io.quarkus.platform:quarkus-amazon-services-bom-quarkus-platform-descriptor
```